### ISSUE TITLE
Puma shows stacktrace in production

### ISSUE THREAD

**Issue by LouisKottmann** [puma/puma#1722]
### Steps to reproduce

I did set:

```
irb(main):002:0> ENV['RAILS_ENV']
=> "production"
irb(main):001:0> Rails.application.config.consider_all_requests_local
=> false
```

Which should avoid the issue.

1) Use curl with an unknown HTTP verb (say, `curl -X WTF https://...`)

### Expected behavior

It should return an error page, or just a 405/501

### Actual behavior

It shows a stacktrace

```
~ $ curl -X WTF https://<my url>
Puma caught this error: WTF, accepted HTTP methods are OPTIONS, GET, HEAD, POST, PUT, DELETE, TRACE, CONNECT, PROPFIND, PROPPATCH, MKCOL, COPY, MOVE, LOCK, UNLOCK, VERSION-CONTROL, REPORT, CHECKOUT, CHECKIN, UNCHECKOUT, MKWORKSPACE, UPDATE, LABEL, MERGE, BASELINE-CONTROL, MKACTIVITY, ORDERPATCH, ACL, SEARCH, MKCALENDAR, and PATCH (ActionController::UnknownHttpMethod)
.../vendor/bundle/ruby/2.4.0/gems/actionpack-4.2.11/lib/action_dispatch/http/request.rb:344:in `check_method'
.../vendor/bundle/ruby/2.4.0/gems/actionpack-4.2.11/lib/action_dispatch/http/request.rb:105:in `request_method'
.../vendor/bundle/ruby/2.4.0/gems/railties-4.2.11/lib/rails/rack/logger.rb:51:in `started_request_message'
.../vendor/bundle/ruby/2.4.0/gems/railties-4.2.11/lib/rails/rack/logger.rb:37:in `block in call_app'
.../ruby/2.4.0/logger.rb:463:in `add'
.../vendor/bundle/ruby/2.4.0/gems/activesupport-4.2.11/lib/active_support/logger.rb:88:in `add'
.../ruby/2.4.0/logger.rb:525:in `info'
.../vendor/bundle/ruby/2.4.0/gems/railties-4.2.11/lib/rails/rack/logger.rb:37:in `call_app'
.../vendor/bundle/ruby/2.4.0/gems/railties-4.2.11/lib/rails/rack/logger.rb:20:in `block in call'
.../vendor/bundle/ruby/2.4.0/gems/activesupport-4.2.11/lib/active_support/tagged_logging.rb:68:in `block in tagged'
.../vendor/bundle/ruby/2.4.0/gems/activesupport-4.2.11/lib/active_support/tagged_logging.rb:26:in `tagged'
.../vendor/bundle/ruby/2.4.0/gems/activesupport-4.2.11/lib/active_support/tagged_logging.rb:68:in `tagged'
.../vendor/bundle/ruby/2.4.0/gems/railties-4.2.11/lib/rails/rack/logger.rb:20:in `call'
.../vendor/bundle/ruby/2.4.0/gems/actionpack-4.2.11/lib/action_dispatch/middleware/request_id.rb:21:in `call'
.../vendor/bundle/ruby/2.4.0/gems/rack-1.6.11/lib/rack/methodoverride.rb:22:in `call'
.../vendor/bundle/ruby/2.4.0/gems/rack-1.6.11/lib/rack/runtime.rb:18:in `call'
.../vendor/bundle/ruby/2.4.0/gems/activesupport-4.2.11/lib/active_support/cache/strategy/local_cache_middleware.rb:28:in `call'
.../vendor/bundle/ruby/2.4.0/gems/actionpack-4.2.11/lib/action_dispatch/middleware/static.rb:120:in `call'
.../vendor/bundle/ruby/2.4.0/gems/rack-1.6.11/lib/rack/sendfile.rb:113:in `call'
.../vendor/bundle/ruby/2.4.0/gems/railties-4.2.11/lib/rails/engine.rb:518:in `call'
.../vendor/bundle/ruby/2.4.0/gems/railties-4.2.11/lib/rails/application.rb:165:in `call'
.../vendor/bundle/ruby/2.4.0/gems/puma-3.12.0/lib/puma/configuration.rb:225:in `call'
.../vendor/bundle/ruby/2.4.0/gems/puma-3.12.0/lib/puma/server.rb:658:in `handle_request'
.../vendor/bundle/ruby/2.4.0/gems/puma-3.12.0/lib/puma/server.rb:472:in `process_client'
.../vendor/bundle/ruby/2.4.0/gems/puma-3.12.0/lib/puma/server.rb:332:in `block in run'
.../vendor/bundle/ruby/2.4.0/gems/puma-3.12.0/lib/puma/thread_pool.rb:133:in `block in spawn_thread'```

### System configuration

**Ruby version**: 2.4.3p205
**Rails version**: 4.2.11
**Puma version**: 3.12.0

**Comment by mic-kul** [comment]
Cannot reproduce on Rails 5.2.2:

```
❯ curl -X WTF http://localhost:3003
* Connected to localhost (127.0.0.1) port 3003 (#0)
> WTF / HTTP/1.1
> Host: localhost:3003
> User-Agent: curl/7.54.0
> Accept: */*
>
< HTTP/1.1 500 Internal Server Error
< Content-Length: 77
<
An unhandled lowlevel error occurred. The application logs may have details.
* Connection #0 to host localhost left intact
```

Stack is printed to STDOUT in Rails process, not to the client:

```
❯ RAILS_ENV=production rails s -p 3003
=> Booting Puma
=> Rails 5.2.2 application starting in production
=> Run `rails server -h` for more startup options
WARNING: Nokogiri was built against LibXML version 2.9.8, but has dynamically loaded 2.9.4
Puma starting in single mode...
* Version 3.12.0 (ruby 2.5.1-p57), codename: Llamas in Pajamas
* Min threads: 5, max threads: 5
* Environment: production
* Listening on tcp://0.0.0.0:3003
Use Ctrl-C to stop
2019-02-23 14:35:56 +0100: Rack app error handling request { WTF / }
#<ActionController::UnknownHttpMethod: WTF, accepted HTTP methods are OPTIONS, GET, HEAD, POST, PUT, DELETE, TRACE, CONNECT, PROPFIND, PROPPATCH, MKCOL, COPY, MOVE, LOCK, UNLOCK, VERSION-CONTROL, REPORT, CHECKOUT, CHECKIN, UNCHECKOUT, MKWORKSPACE, UPDATE, LABEL, MERGE, BASELINE-CONTROL, MKACTIVITY, ORDERPATCH, ACL, SEARCH, MKCALENDAR, and PATCH>
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/actionpack-5.2.2/lib/action_dispatch/http/request.rb:426:in `check_method'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/actionpack-5.2.2/lib/action_dispatch/http/request.rb:135:in `request_method'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/railties-5.2.2/lib/rails/rack/logger.rb:51:in `started_request_message'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/railties-5.2.2/lib/rails/rack/logger.rb:37:in `block in call_app'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/2.5.0/logger.rb:465:in `add'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activesupport-5.2.2/lib/active_support/logger.rb:89:in `add'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/2.5.0/logger.rb:527:in `info'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/railties-5.2.2/lib/rails/rack/logger.rb:37:in `call_app'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/railties-5.2.2/lib/rails/rack/logger.rb:26:in `block in call'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activesupport-5.2.2/lib/active_support/tagged_logging.rb:71:in `block in tagged'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activesupport-5.2.2/lib/active_support/tagged_logging.rb:28:in `tagged'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activesupport-5.2.2/lib/active_support/tagged_logging.rb:71:in `tagged'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/railties-5.2.2/lib/rails/rack/logger.rb:26:in `call'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/actionpack-5.2.2/lib/action_dispatch/middleware/remote_ip.rb:81:in `call'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/actionpack-5.2.2/lib/action_dispatch/middleware/request_id.rb:27:in `call'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/rack-2.0.6/lib/rack/runtime.rb:22:in `call'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activesupport-5.2.2/lib/active_support/cache/strategy/local_cache_middleware.rb:29:in `call'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/actionpack-5.2.2/lib/action_dispatch/middleware/executor.rb:14:in `call'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/rack-2.0.6/lib/rack/sendfile.rb:111:in `call'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/railties-5.2.2/lib/rails/engine.rb:524:in `call'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/puma-3.12.0/lib/puma/configuration.rb:225:in `call'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/puma-3.12.0/lib/puma/server.rb:665:in `handle_request'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/puma-3.12.0/lib/puma/server.rb:479:in `process_client'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/puma-3.12.0/lib/puma/server.rb:338:in `block in run'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/puma-3.12.0/lib/puma/thread_pool.rb:133:in `block in spawn_thread'
2019-02-23 14:36:04 +0100: Rack app error handling request { WTF / }
#<ActionController::UnknownHttpMethod: WTF, accepted HTTP methods are OPTIONS, GET, HEAD, POST, PUT, DELETE, TRACE, CONNECT, PROPFIND, PROPPATCH, MKCOL, COPY, MOVE, LOCK, UNLOCK, VERSION-CONTROL, REPORT, CHECKOUT, CHECKIN, UNCHECKOUT, MKWORKSPACE, UPDATE, LABEL, MERGE, BASELINE-CONTROL, MKACTIVITY, ORDERPATCH, ACL, SEARCH, MKCALENDAR, and PATCH>
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/actionpack-5.2.2/lib/action_dispatch/http/request.rb:426:in `check_method'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/actionpack-5.2.2/lib/action_dispatch/http/request.rb:135:in `request_method'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/railties-5.2.2/lib/rails/rack/logger.rb:51:in `started_request_message'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/railties-5.2.2/lib/rails/rack/logger.rb:37:in `block in call_app'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/2.5.0/logger.rb:465:in `add'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activesupport-5.2.2/lib/active_support/logger.rb:89:in `add'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/2.5.0/logger.rb:527:in `info'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/railties-5.2.2/lib/rails/rack/logger.rb:37:in `call_app'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/railties-5.2.2/lib/rails/rack/logger.rb:26:in `block in call'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activesupport-5.2.2/lib/active_support/tagged_logging.rb:71:in `block in tagged'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activesupport-5.2.2/lib/active_support/tagged_logging.rb:28:in `tagged'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activesupport-5.2.2/lib/active_support/tagged_logging.rb:71:in `tagged'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/railties-5.2.2/lib/rails/rack/logger.rb:26:in `call'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/actionpack-5.2.2/lib/action_dispatch/middleware/remote_ip.rb:81:in `call'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/actionpack-5.2.2/lib/action_dispatch/middleware/request_id.rb:27:in `call'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/rack-2.0.6/lib/rack/runtime.rb:22:in `call'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activesupport-5.2.2/lib/active_support/cache/strategy/local_cache_middleware.rb:29:in `call'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/actionpack-5.2.2/lib/action_dispatch/middleware/executor.rb:14:in `call'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/rack-2.0.6/lib/rack/sendfile.rb:111:in `call'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/railties-5.2.2/lib/rails/engine.rb:524:in `call'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/puma-3.12.0/lib/puma/configuration.rb:225:in `call'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/puma-3.12.0/lib/puma/server.rb:665:in `handle_request'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/puma-3.12.0/lib/puma/server.rb:479:in `process_client'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/puma-3.12.0/lib/puma/server.rb:338:in `block in run'
/Users/mic-kul/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/puma-3.12.0/lib/puma/thread_pool.rb:133:in `block in spawn_thread'
```

**Comment by evanphx** [comment]
Yes, that output appears to be on puma's stdout, not in the response.

**Comment by tokhi** [comment]
Is this issue resolved? can't see any feedback.

**Comment by LouisKottmann** [comment]
From memory, I ended up screening calls on the nginx frontend with a whitelist.

**Comment by nateberkopec** [comment]
This issue is resolved on Puma 3.12+.
